llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,371 @@
+
+  Register getAssignedPhysReg(const MachineOperand &MO) const;
+};
+
+Register AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register PhysReg = MO.getReg();
+  if (PhysReg.isVirtual())
+    PhysReg = VRM->getPhys(PhysReg);
+  return PhysReg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+bool hasSubRegCopy(
+    const unsigned ID,
+    const std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+        &SubRegDefs) {
+  if (SubRegDefs.find(ID) == SubRegDefs.end())
+    return false;
+
+  for (const MachineInstr *MI : SubRegDefs.at(ID)) {
+    if (MI->isCopyLike())
+      return true;
+  }
+  return false;
+}
+
+std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+getAllSubRegDefs(const MachineBasicBlock &MBB) {
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      unsigned ID = Def.getReg().id();
+
+      if (Def.isUndef())
+        SubRegDefs[ID] = SmallVector<const MachineInstr *, 4>();
+
+      if (SubRegDefs.find(ID) == SubRegDefs.end())
+        continue;
+
+      SubRegDefs[ID].push_back(&MI);
+    }
+  }
+  return SubRegDefs;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  bool Modified = false;
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  BitVector UsedPhyRegs(TRI->getNumRegs());
+  SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+  SmallSet<MCPhysReg, 16> ForbiddenPhysRegs = getMBBForbiddenReplacements(MBB);
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs =
+      getAllSubRegDefs(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+        continue;
+      
+      MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+      if (isWAWCandidate(Reg, MI, UsedPhyRegs, SubRegDefs) &&
```

Comment:
Nit: I feel this check is more about profitability, compared to the checks above that are about "legality". Maybe we can rename that function to `isWorthRenaming()`?

Reply:
changed the name